core/src/main/scala/cats/data/Chain.scala

```diff
@@ -565,33 +565,19 @@ sealed abstract class Chain[+A] {
   /**
    * Returns the number of elements in this structure
    */
-  final def length: Long = {
-    // TODO: consider optimizing for `Chain.Wrap` case.
-    //       Some underlying seq may not need enumerating all elements to calculate its size.
-    val iter = iterator
-    var i: Long = 0
-    while (iter.hasNext) { i += 1; iter.next(); }
-    i
-  }
+  final def length: Long =
+    this match {
+      case Empty                   => 0
+      case Wrap(seq)               => seq.length.toLong
+      case Singleton(a)            => 1
+      case Append(leftNE, rightNE) => leftNE.length + rightNE.length
```

Comment:
This case for `Append` triggers recursive evaluation of `length` (if the `Chain` consist of a bunch of `Append`-s). It may not be stack safe. IMO it would be better to keep the old iteration based evaluation for the `Append` case.

Reply:
... on the second thought (and as an alternative to just iteration), we can consider implementing a separate (perhaps private) method based on `Eval` structure:
```
def lengthEval: Eval[Long]
```
This method can handle all the length calculations in a stack-safe way.
Then the regular `length` can call that method to trigger all the calculation.

Not sure if it is really worth it, though...